src/Hyphenator/HyphenationRule.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Hyphenator;
+
+class HyphenationRule
+{
+    private string $pattern;
+    private array $levels;
+
+    public function __construct(string $rule)
+    {
+        $validNumbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
+        $this->levels = [];
+        $chars = str_split($rule);
+        if (strlen($rule) > 0 && $this->thisCharValid(0, $chars)) {
+            $this->levels[] = 0;
+        }
+        for($i = 0; $i < count($chars); $i++) {
+            if ($this->thisCharValid($i, $chars) && $this->nextCharValid($i, $chars)) {
+                $this->levels[] = 0;
+            } elseif (is_numeric($chars[$i])) {
+                $this->levels[] = (int) $chars[$i];
+            }
+        }
+        $this->pattern = str_replace($validNumbers, '', $rule);
+    }
+
+    public function getRule(): string
+    {
+        return $this->pattern;
+    }
+
+    public function getRuleLength(): int
+    {
+        return strlen($this->pattern);
+    }
+
+    public function getLevels(): array
+    {
+        return $this->levels;
+    }
+
+    public function getLevelsLength() :int
+    {
+        return count($this->levels);
+    }
+
+    public function matchesStart(HyphenationWord $word): bool
+    {
+        $wordPattern = '.' . strtolower($word->getWord()) . '.';
+        $wordLevelsLength = $word->getLevelsLength();
+        $ruleLevelsLength = count($this->levels);
+        $startsWithPattern = str_starts_with($wordPattern, $this->pattern);
+
+        return $startsWithPattern
+            && (
+                $ruleLevelsLength < $wordLevelsLength + 1
+                || $ruleLevelsLength === $wordLevelsLength + 1
+                && $this->levels[$ruleLevelsLength - 1] === 0
+            );
+    }
+
+    public function matchesEnd(HyphenationWord $word): bool
+    {
+        $wordPattern = '.' . strtolower($word->getWord()) . '.';
+        $wordLevelsLength = $word->getLevelsLength();
+        $ruleLevelsLength = count($this->levels);
+        $endsWithPattern = str_ends_with($wordPattern, $this->pattern);
+
+        return $endsWithPattern
+            && (
+                $ruleLevelsLength < $wordLevelsLength + 1
+                || $ruleLevelsLength === $wordLevelsLength + 1
+                && $this->levels[0] === 0
+            );
+    }
+
+    public function matchesMiddle(HyphenationWord $word): int
```

Comment:
previous two `matches*` methods returns booleands this one returns int, for consistency it would be better for it to return boolean, or maybe it should be renamed to `compareMiddle`?